### PR TITLE
Properly chunk text to satisfy desired chunk size when adding extra_info to index node

### PR DESCRIPTION
**Pull request by Disiok**
### Summary
* This fixes the bug where text chunking logic in index construction ignores the `extra_info` metadata that gets prepended to the document text chunks. 

### Detailed Changes
* Extend `TokenTextSplitter.split_text` to support additional optional parameter  `extra_info_str`
> This allows it to explicitly consider extra info str that will be added to the chunk at query time, which reduces the effective chunk size that we can have
* Modify `BaseGPTIndex._get_nodes_from_document` and `GPTTreeIndexBuilder._get_nodes_from_document` to call `split_text` with `extra_info_str` when constructing nodes with `extra_info`
> We also pull out `get_nodes_from_document` into helper function for ease of unit testing

### Testing
Unit tests

Closes GPT-16

### CODE REVIEW

**Review by jerryjliu — COMMENTED** [review]
high-level this looks reasonable. a few questions/comments:

1) why put extra_info_str on the Document object instead of the Node object?
2) We recently landed a change to fix how we compute start/end indices for each node (given overlaps), which changed the usage of split_text to `split_text_with_overlaps`: https://github.com/jerryjliu/gpt_index/pull/297/files. You should repull main

**Comment by Disiok** [comment]
> high-level this looks reasonable. a few questions/comments:
> 
> 1. why put extra_info_str on the Document object instead of the Node object?
> 2. We recently landed a change to fix how we compute start/end indices for each node (given overlaps), which changed the usage of split_text to `split_text_with_overlaps`: https://github.com/jerryjliu/gpt_index/pull/297/files. You should repull main

1. `_get_nodes_from_document` takes `BaseDocument` as input, which I assume could be either `Document` or `Node`. So it made sense to me to add `extra_info_str` property at `BaseDocument` level, instead of `Node` level.
> Right now, an inconsistency that exists is `Document.get_text` does not return extra_info, whereas `Node` returns `extra_info` if it's not None. Not sure if this is intention, but seems possibly bug prone.
2. Pulled and updated
> I'm confused as to why `GPTTreeIndexBuilder._get_nodes_from_document` is not updated to use `split_text_with_overlaps` instead of `split_text`

**Comment by jerryjliu** [comment]
> > Right now, an inconsistency that exists is `Document.get_text` does not return extra_info, whereas `Node` returns `extra_info` if it's not None. Not sure if this is intention, but seems possibly bug prone.

The reason for that was mostly to inject metadata into each text chunk (which is what a Node is). You don't want to inject the metadata for each Document. 

> 2. Pulled and updated
> 
> > I'm confused as to why `GPTTreeIndexBuilder._get_nodes_from_document` is not updated to use `split_text_with_overlaps` instead of `split_text`

Hmm maybe that was missed

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,3 @@
-.vscode/
```
Comment (jerryjliu) [inline .gitignore:1]: given #308 can you unset these changes?

File: gpt_index/langchain_helpers/text_splitter.py [hunk 2]
```diff
@@ -103,15 +103,32 @@ def _process_splits(self, splits: List[str]) -> List[str]:
...
+    ) -> List[TextSplit]:
         """Split incoming text and return chunks with overlap size."""
         if text == "":
             return []
+
+        # NOTE: Consider extra info str that will be added to the chunk at query time
+        #       This reduces the effective chunk size that we can have
+        if extra_info_str is not None:
+            # NOTE: extra 2 newline chars for formatting when prepending in query
+            num_extra_tokens = len(self.tokenizer(f"{extra_info_str}\n\n")) + 1
```
Comment (jerryjliu) [inline gpt_index/langchain_helpers/text_splitter.py:122]: out of curiosity: why + 1?

**Review by jerryjliu — APPROVED** [review]
approving, though I have some comments

File: .gitignore (hunk 1 above)
Comment (Disiok) [reply .gitignore:1]: oh hmm this leaked in oops.

File: gpt_index/langchain_helpers/text_splitter.py (hunk 2 above)
Comment (Disiok) [reply gpt_index/langchain_helpers/text_splitter.py:122]: seems like e.g.  `apple\n\n` is 3 tokens, but `apple\n\napple` is 5 tokens. I'm not familiar enough with tokenization to know why 😂

Comment (jerryjliu) [reply gpt_index/langchain_helpers/text_splitter.py:122]: Hm ok. can you leave a TODO? the tokenizer is a bit of a black box. i imagine if there's a token exceeded bug later we may have to dig into this a bit more, but maybe we don't have to worry for now

Comment (jerryjliu) [reply gpt_index/langchain_helpers/text_splitter.py:122]: going to merge this for now to unblock

**Pull request merged**